crates/circuit/src/dag_node.rs

```diff
@@ -0,0 +1,329 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::circuit_instruction::CircuitInstruction;
+use pyo3::prelude::*;
+use pyo3::sync::GILOnceCell;
+use pyo3::types::{PyDict, PyList, PySequence, PyString, PyTuple};
+use pyo3::{intern, PyObject, PyResult};
+
+static SEMANTIC_EQ_PYTHON: GILOnceCell<PyResult<PyObject>> = GILOnceCell::new();
+
+/// Parent class for DAGOpNode, DAGInNode, and DAGOutNode.
+#[pyclass(module = "qiskit._accelerate.circuit", subclass)]
+#[derive(Clone, Debug)]
+pub struct DAGNode {
+    #[pyo3(get, set)]
+    pub _node_id: isize,
+}
+
+#[pymethods]
+impl DAGNode {
+    #[new]
+    #[pyo3(signature=(nid=-1))]
+    fn new(nid: isize) -> Self {
+        DAGNode { _node_id: nid }
+    }
+
+    fn __getstate__(&self) -> isize {
+        self._node_id
+    }
+
+    fn __setstate__(&mut self, nid: isize) {
+        self._node_id = nid;
+    }
+
+    fn __lt__(&self, other: &DAGNode) -> bool {
+        self._node_id < other._node_id
+    }
+
+    fn __gt__(&self, other: &DAGNode) -> bool {
+        self._node_id > other._node_id
+    }
+
+    fn __str__(_self: &Bound<DAGNode>) -> String {
+        format!("{}", _self.as_ptr() as usize)
+    }
+
+    fn __hash__(&self, py: Python) -> PyResult<isize> {
+        self._node_id.into_py(py).bind(py).hash()
+    }
+
+    /// Check if DAG nodes are considered equivalent, e.g., as a node_match for
+    /// :func:`rustworkx.is_isomorphic_node_match`.
+    ///
+    /// Args:
+    ///     node1 (DAGOpNode, DAGInNode, DAGOutNode): A node to compare.
+    ///     node2 (DAGOpNode, DAGInNode, DAGOutNode): The other node to compare.
+    ///     bit_indices1 (dict): Dictionary mapping Bit instances to their index
+    ///         within the circuit containing node1
+    ///     bit_indices2 (dict): Dictionary mapping Bit instances to their index
+    ///         within the circuit containing node2
+    ///
+    /// Return:
+    ///     Bool: If node1 == node2
+    #[staticmethod]
+    #[pyo3(signature = (*args), text_signature = "(node1, node2, bit_indices1, bit_indices2)")]
+    fn semantic_eq<'py>(py: Python<'py>, args: &Bound<PyTuple>) -> PyResult<Bound<'py, PyAny>> {
+        let semantic_eq_fn = SEMANTIC_EQ_PYTHON
+            .get_or_init(py, || {
+                let module = PyModule::import_bound(py, "qiskit.dagcircuit.dagnode")?;
+                Ok(module.getattr("_semantic_eq")?.unbind())
+            })
+            .as_ref()
+            .map_err(|e| e.clone_ref(py))?;
+        semantic_eq_fn.bind(py).call1(args)
+    }
```

Comment:
Possibly could we make `semantic_eq` a `#[classattr]` that takes only `py` and returns `semantic_eq_fn`? That way we don't need to apply a docstring or keep the args in sync - we'd just be attaching the outer function as a `staticmethod`, rather than nesting Python calls.

Reply:
To avoid the import and indirection, I pulled all of this out of Rust and instead just assign the static function as a class attribute of `DAGNode` within `dagnode.py`.

041f3419b06b6f3a8c1a9cb0cc1db0021d357d7c

Any objections to that approach?

Reply:
Nah, that seems fine to me for now.  One day long in the future that'll want to be a Rust-space method, but for now that seems like a good way of doing things.